third_party/parser/include/ruby_parser/lexer.hh

```diff
@@ -148,8 +148,16 @@ public:
 
     lexer(diagnostics_t &diag, ruby_version version, const std::string &source_buffer_);
 
+    // Main interface consumed by yylex function in parser
     token_t advance();
 
+    // Useful for error recovery. Manually pushes `tok_to_push` (usually: the parser's current
+    // lookahead token) back onto the front of the token_queue, and then allocates a new token and
+    // returns it.
+    token_t unadvance(token_t tok_to_push, token_type type, size_t start, size_t end, const std::string &str);
+
+    std::string_view tok_view(size_t start, size_t end) const;
```

Comment:
Mild preference for calling this `tok_view_from_indices`, though something shorter would be nice.